src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+    }
+  };
+
+  const handleAnd = (expr: BinaryExpr): PatternDocument => {
+    const left = handleExpr(expr.left);
+    const right = handleExpr(expr.right);
+
+    return andDocuments(left, right);
+  };
+
+  const andDocuments = (
+    classDocument1: PatternDocument,
+    classDocument2: PatternDocument
+  ): PatternDocument => {
+    const allKeys = [
+      ...new Set([
+        ...Object.keys(classDocument1.doc),
+        ...Object.keys(classDocument2.doc),
+      ]),
+    ];
+
+    return allKeys.reduce((doc, key) => {
+      // if key is only in one document, return it
+      const pattern = !(key in classDocument1.doc)
+        ? classDocument2.doc?.[key]
+        : !(key in classDocument2.doc)
+        ? classDocument1.doc?.[key]
+        : undefined;
+      // if the key is in both documents, try to merge them
+      const mergedPattern =
+        pattern ??
+        andMergePattern(key, classDocument1.doc[key], classDocument2.doc[key]);
+      return {
+        doc: {
+          ...doc.doc,
+          [key]: mergedPattern,
+        },
+      };
+    }, {} as PatternDocument);
+  };
+
+  const andMergePattern = (
+    key: string,
+    pattern1: PatternDocument | Pattern,
+    pattern2: PatternDocument | Pattern
+  ): PatternDocument | Pattern => {
+    // both are documents, lets merge the documents
+    if (isPatternDocument(pattern1) && isPatternDocument(pattern2)) {
+      return andDocuments(pattern1, pattern2);
+      // both are patterns, merge the patterns
+    } else if (!isPatternDocument(pattern1) && !isPatternDocument(pattern2)) {
+      if (isNumericAggregationPattern(pattern1)) {
+        if (isNumericAggregationPattern(pattern2)) {
+          return andNumericAggregation(pattern1, pattern2);
+        }
+        if (isNumericRangePattern(pattern2)) {
+          return andNumericAggregationWithRange(pattern1, pattern2);
+        }
+      }
+      if (isNumericAggregationPattern(pattern2)) {
+        if (isNumericRangePattern(pattern1)) {
+          return andNumericAggregationWithRange(pattern2, pattern1);
+        }
+      }
+      if (isNumericRangePattern(pattern1) && isNumericRangePattern(pattern2)) {
+        return joinNumericRange(pattern1, pattern2);
+      }
+      // TODO: expand the supported cases
+      throw new Error(
+        `Cannot apply AND to patterns ${JSON.stringify(
+          pattern1
+        )} and ${JSON.stringify(pattern2)}`
+      );
+    }
+    // we don't know how to do this (yet?)
+    throw new Error(`Patterns of key ${key} defined at different levels.`);
+  };
+
+  /**
+   * OR Boolean logic handler.
+   *
+   * Resolves the left and right branches and then tries to merge together the results of each.
+   */
+  const handleOr = (expr: BinaryExpr): PatternDocument => {
+    const left = handleExpr(expr.left);
+    const right = handleExpr(expr.right);
+
+    return orDocuments(left, right);
+  };
+
+  /**
+   * Event bridge does not support OR across fields, only AND logic
```

Comment:
Much wow! :/

Reply:
Is there a technical reason or is EB dumb?

Reply:
```ts
{
   field1: ["value"],
   field2: ["value"]
}
```

this represents `&&`  between field1 and field2. There is no way to do a `||` without creating multiple rules that are mutually exclusive

```ts
// one
{
   field1: [{ anything-but: "value"}],
   field2: ["value"]
}

// other
{
   field1: ["value"],
   field2: [{ anything-but: "value"}]
}

// both
{
   field1: ["value"],
   field2: ["value"]
}
```

and XOR

```ts
// one
{
   field1: [{ anything-but: "value"}],
   field2: ["value"]
}

// other
{
   field1: ["value"],
   field2: [{ anything-but: "value"}]
}
```

Reply:
and multiple rules that point to the same target would generate multiple events, which is why the rules would need to stay perfectly mutually exclusive. 

This, pointing to the same target, will likely send duplicate events when both conditions are met.

```ts
// one
{
   field2: ["value"]
}

// other
{
   field1: ["value"],
}
```